lib/api/dsl/storage/index.js

```diff
@@ -266,6 +266,124 @@ class Storage {
+    const tt = this.testTables[index][collection];
+
+    if (tt.removedFilters[filter.id]) {
+      return;
+    }
+
+    tt.removedFilters[filter.id] = true;
+    tt.removedFiltersCount++;
+
+    if (tt.removedFiltersCount === tt.filtersCount) {
+      if (containsOne(this.testTables[index])) {
+        delete this.testTables[index];
+      }
+      else {
+        delete this.testTables[index][collection];
+      }
+
+      return;
+    }
+
+    // Declaring "i" inside the "for" statement downgrades
+    // performances by a factor of 3 to 4
+    // Should be fixed in later V8 versions
+    // (tested on Node 6.9.x)
+    let i; // NOSONAR
+
+    for(i = 0; i < filter.subfilters.length; i++) {
+      if (filter.subfilters[i].filters.length === 1) {
+        tt.removedConditions.insert(filter.subfilters[i].cidx);
+      }
+    }
+
+    /*
+     Perform a reindex only if the number of deleted conditions is greater
+     than 10% of the total number of registered conditions
+
+     In that case, we flag the reindex operation and launch it a few seconds
+     after that. This allows large unsubscriptions activity to finish.
+     Best case scenario: the test table is removed altogether, thus avoiding reindexation.
+     Worst case scenario: we still have a reindexation to perform, but luckily
+     including a lot more unsubscriptions to remove, greatly reducing the overall reindex cost
+     */
+    if (!tt.reindexing && tt.removedConditions.array.length > tt.clength / 10) {
+      tt.reindexing = true;
+      setTimeout(() => this._reindexTestTable(index, collection), 5000);
+    }
+  }
+
+  /**
+   * Performs a reindexation of a test table
+   *
+   * @param {string} index
+   * @param {string} collection
+   */
+  _reindexTestTable (index, collection) {
+    /*
+     * Since the reindexation has been triggered, the test table might have been
+     * destroyed
+     */
+    if (!this.testTables[index] || !this.testTables[index][collection] || !this.testTables[index][collection].reindexing) {
+      return;
+    }
+
+    const tt = this.testTables[index][collection];
+
+    /*
+     * If there is a huge increase in new subscriptions, we should postpone the
+     * reindexation to a later time.
+     */
+    if (tt.removedConditions.array.length < tt.clength / 10) {
+      return;
+    }
+
+    // rebuild conditions index
+    const
+      conditionKeys = Object.keys(this.subfilters[index][collection]),
+      conditions = new Uint8Array(tt.conditions.length);
+
+    for (let i = 0; i < conditionKeys.length; i++) {
+      const
+        key = conditionKeys[i],
+        sf = this.subfilters[index][collection][key];
+
+      conditions[i] = sf.conditions.length;
+      sf.cidx = i;
+    }
+    tt.conditions = conditions;
+    tt.clength = conditionKeys.length;
+
+    // rebuild filter index pointers
+    if (tt.removedFiltersCount) {
+      for (let i = 0; i < Object.keys(this.filtersIndex[index][collection]).length; i++) {
```

Comment:
Unfortunately yes: unregistration and reindexation are both very time-consuming tasks, and these therefore should be considered critical code.